### PR TITLE
Observability To Colocated Auction Runloop

### PR DESCRIPTION
**Pull request by nlordell**
# Description

This PR refactors the autopilot runloop to add some additional observability to the solver competition. This will allow us to have a better idea of how individual solvers are behaving.

As a note on the code and how to record metrics - I created methods on the `Metrics` type in order to increment/observe things. I'm not 100% happy with this, but wanted to see what others think in terms of legibility. 

# Changes

- [x] Added gauge for last auction ID seen in the runloop.
- [x] Specific error types for all driver interactions
- [x] Added metrics counters for results of driver interactions (solving, revealing and settling), and specifically a histogram for `solve` timings.

## How to test

The code should be covered by existing E2E tests in the services.

### CODE REVIEW

File: crates/autopilot/src/run_loop.rs
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:388]: Interesting. Are we putting the auction ID data also on-chain?

File: crates/autopilot/src/run_loop.rs
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:617]:
```suggestion
#[error("the solver proposed a 0-score solution")]
```

**Review by fleupold — APPROVED** [review]
> I created methods on the Metrics type in order to increment/observe things. I'm not 100% happy with this, but wanted to see what others think in terms of legibility.

Why are you not happy with this? Seems fine and readable to me...

File: crates/autopilot/src/run_loop.rs
Comment (nlordell) [reply crates/autopilot/src/run_loop.rs:388]: Yes, we started doing this.

File: crates/autopilot/src/run_loop.rs
Comment (nlordell) [reply crates/autopilot/src/run_loop.rs:478]: So, my understanding is that this is related to where the compiler places the `Drop`. For example, see [this issue](https://github.com/rust-lang/rust/issues/37612).

That being said, I _don't think_ it was a problem in our usage here, I just adapted the existing unbounded `loop` to be a `while` loop and just reworded the comment that was already there (I assumed it was done this way for a reason, I didn't verify it as such):

https://github.com/cowprotocol/services/blob/4dbeeb8170c52fe76b02f8730d30064f3426ad35/crates/autopilot/src/run_loop.rs#L435-L441

**Comment by nlordell** [comment]
> Why are you not happy with this? Seems fine and readable to me...

I feel like the `Metrics::get().observe(...)` is peppered in the code and slightly inconsistent, and adds cognitive load when reading (instead of having a single place where we record all metrics for example).

**Comment by fleupold** [comment]
>I feel like the Metrics::get().observe(...) is peppered in the code and slightly inconsistent

In the new architecture we would define a separate observe module and do all logging/metrics there? I kind of like that as well but is probably a large refactor for the autopilot...

File: crates/autopilot/src/run_loop.rs
Comment (MartinquaXD) [inline crates/autopilot/src/run_loop.rs:313]: nit: Is there a reason the current time needs to be fetched for each solver individually. I think there is no meaningful loss in precision if we were to fetch the time once before building the futures.

File: crates/autopilot/src/run_loop.rs
Comment (MartinquaXD) [inline crates/autopilot/src/run_loop.rs:322]: This metric doesn't seem useful since it will always report the exact same solving time for all solvers because you use `join_all` which will only return when the last solvers finishes.
If you want to fetch individual timings you can collect the futures into `FuturesUnordered` which behaves like a stream and yields the results exactly when they become available.

Alternatively you could move most of the logic currently inside `.fold()` into the future prepared in `join_all()`. That would be the easiest to read, gives accurate timings and can continue to use `join_all()`.

**Comment by nlordell** [comment]
> In the new architecture we would define a separate observe module and do all logging/metrics there?

My comment is unrelated to this, and more in the context of this individual module:

> we update metrics in `single_auction` and in `competition` methods rather arbitrarily, instead of in a single function of the `RunLoop` type. The metrics adds additional control flow at those points, and it would be nice for it to be more consolidated.

And not about the metrics architecture in the `driver`. I don't think either @MartinquaXD nor myself are particularly fond of logging from a single module (there are some downsides) so I would advise against it (and generally change how we do things in the driver to be more consistent with the rest of the project; I personally see it as an unsuccessful experiment).

File: crates/autopilot/src/run_loop.rs
Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs:478]: I think we should review whether this is actually needed (I also think it's not in this case) just to avoid further confusion in the future.
I'm fine with leaving this as is since this PR is urgent.

**Review by MartinquaXD — APPROVED** [review]
Nice observability improvement! Just seems to me that we would currently not get any meaningful timing information out of the driver runs due to `join_all()`.
Approving assuming this gets addressed.

**Comment by MartinquaXD** [comment]
> I feel like the Metrics::get().observe(...) is peppered in the code and slightly inconsistent

IMO we could also move the `Metrics::get()` part into each `.observe()` function implementation. Then it's just `Metrics::observe()` which seems slightly nicer IMO. No strong opinion, though.

File: crates/autopilot/src/run_loop.rs
Comment (nlordell) [reply crates/autopilot/src/run_loop.rs:322]: Oops!

Comment (nlordell) [reply crates/autopilot/src/run_loop.rs:322]: I changed it to have the futures do `start.elapsed` instead of passing in the `Instant`.

File: crates/autopilot/src/run_loop.rs
Comment (nlordell) [reply crates/autopilot/src/run_loop.rs]: d84b7dc3d95bb7aeb93b7940106a6174e71d69b6

File: crates/autopilot/src/run_loop.rs
Comment (nlordell) [reply crates/autopilot/src/run_loop.rs]: d84b7dc3d95bb7aeb93b7940106a6174e71d69b6

**Pull request merged**
